bpf/lib/nodeport.h

```diff
@@ -1438,6 +1525,25 @@ nodeport_extract_dsr_v4(struct __ctx_buff *ctx, struct iphdr *ip4,
 		}
 	}
 
+#if defined(IS_BPF_OVERLAY) && \
+	defined(ENABLE_DSR) && DSR_ENCAP_MODE == DSR_ENCAP_GENEVE
+	{
+		struct geneve_opt4 gopt;
+		int ret = 0;
+
+		ret = ctx_get_tunnel_opt(ctx, &gopt, sizeof(gopt));
+
+		if (ret > 0) {
+			if (bpf_ntohs(gopt.opt_class) == DSR_GENEVE_OPT_CLASS &&
+			    gopt.type == DSR_GENEVE_OPT_TYPE) {
+				*dsr = true;
+				*port = gopt.port;
+				*addr = gopt.addr;
+				return 0;
+			}
+		}
+	}
+#else
```

Comment:
With this we would still walk through the TCP-specific code above and below, right? So I would take this even further - only call `nodeport_extract_dsr_v4()` iff

1. `DSR_ENCAP_MODE == GENEVE` && `IS_BPF_OVERLAY`, or
2. `DSR_ENCAP_MODE != GENEVE` && !`IS_BPF_OVERLAY`

And then have a simpler condition in here - if `defined(IS_BPF_OVERLAY)` look for the tunnel-opt, else look for IP-Option.